johnnydep/compat.py

```diff
@@ -1,17 +1,42 @@
 import oyaml
 
-
 try:
-    from urllib.request import urlretrieve
+    from urllib.parse import quote, urlparse
+    from urllib import request as urllib2
 except ImportError:
     # Python 2
-    from urllib import urlretrieve
+    import urllib2
+    from urllib2 import quote
+    from urlparse import urlparse
+
+
+def urlretrieve(url, filename, data=None, auth=None):
+    if auth is not None:
```

Comment:
The test coverage is not actually getting into this branch, because this function was mocked out entirely in your tests.

Reply:
how do you propose to test `urlretrieve()`? I'm no expert at pytest, I always try to keep tests completely offline but not sure what's best practice in this specific case.

the test i wrote covers pretty much all the cases how this function can be called with/without an auth keyword. also, the auth part is copied from official py docs. would it not be overkill to test that specific part with an actual api running locally?

Reply:
Yes completely offline is best, but I still like to aim for 100% coverage of the project code (whether it's copied from docs or not).  That means mocking at the `opener.open` line, I think.